### PR TITLE
feat(java_common): implement standard compilation unit digesting

### PR DESCRIPTION
**Pull request by jaysachs**
Not hooked in anywhere.

Part of #3680

### CODE REVIEW

File: kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import com.google.devtools.kythe.proto.Analysis.CompilationUnit.FileInput;
+import com.google.devtools.kythe.proto.Storage.VName;
+import com.google.protobuf.Any;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/** Utilities for working with CompilationUnit protos. */
```
Comment (shahms) [inline kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java]: Do we want to document the "standard" digesting independent of a particular language first?

Comment (jaysachs) [reply kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java]: Done.

File: kythe/javatests/com/google/devtools/kythe/platform/shared/CompilationUnitsTest.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import com.google.devtools.kythe.proto.Analysis.FileInfo;
+import com.google.devtools.kythe.proto.Storage.VName;
+import com.google.protobuf.Any;
+import com.google.protobuf.ByteString;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class CompilationUnitsTest {
```
Comment (schroederc) [inline kythe/javatests/com/google/devtools/kythe/platform/shared/CompilationUnitsTest.java]: The tests should share testdata across the languages to enforce compatibility.

Comment (jaysachs) [reply kythe/javatests/com/google/devtools/kythe/platform/shared/CompilationUnitsTest.java]: Done.

File: kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  private static final byte[] TAG_DET = "DET".getBytes(CHARSET);
+  private static final byte[] TAG_ENV = "ENV".getBytes(CHARSET);
+
+  /**
+   * Computes the standard digest for the specified {@link CompilationUnit} in stringified
+   * hexidecimal form.
+   */
+  public static String digestFor(CompilationUnit compilationUnit) {
+    CompilationUnit unit = canonicalize(compilationUnit);
+    ByteArrayDataOutput w = ByteStreams.newDataOutput();
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java]: Write directly to `DIGEST.newHasher()`?

File: kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      // If ri[j] ≠ ri[i], it is a new element, not a duplicate.  Move it
+      // next in sequence after i and advance i; this ensures we keep the
+      // first occurrence among duplicates.
+      //
+      // Otherwise, ri[k] == ri[i] for i <= k <= j, i.e., we are scanning a
+      // run of duplicates, and should leave i where it is.
+      if (compareInputs(ri.get(i), ri.get(j)) != 0) {
+        i++;
+        if (i != j) {
+          FileInput tmp = ri.get(i);
```
Comment (schroederc) [inline kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java]: I think you just need `ri.set(i, ri.get(j));` without a swap.  If `i != j`, then `i` was already determined to be a dup and can just be overwritten.

File: kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java (hunk 3 above)
Comment (jaysachs) [reply kythe/java/com/google/devtools/kythe/platform/shared/CompilationUnits.java]: Done

**Pull request merged**